<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Пётр, привет!👋</b> 

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/> 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
 
     
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



# Проект спринта №7 - анализ продаж игр по жанрам и рынкам

- Автор: Поляков П. В.
- Дата: 23.02.2023

### Цели и задачи проекта

<font color='#777778'>Познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных. Проанализировать датасет с информацией о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.</font>

### Описание данных

<font color='#777778'>Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
* Name — название игры.
* Platform — название платформы.
* Year of Release — год выпуска игры.
* Genre — жанр игры.
* NA sales — продажи в Северной Америке (в миллионах проданных копий).
* EU sales — продажи в Европе (в миллионах проданных копий).
* JP sales — продажи в Японии (в миллионах проданных копий).
* Other sales — продажи в других странах (в миллионах проданных копий).
* Critic Score — оценка критиков (от 0 до 100).
* User Score — оценка пользователей (от 0 до 10).
* Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</font>

### Содержимое проекта

<font color='#777778'>
    
- **Проверка названий столбцов:**
  - Вывести названия всех столбцов.
  - Привести все названия к стилю snake_case (нижний регистр, пробелы заменить на подчёркивания).

- **Проверка типов данных:**
  - Проверить корректность типов данных.
  - Преобразовать числовые столбцы, содержащие строковые значения (например, заменить "unknown" на пропуски).

- **Анализ пропусков:**
  - Подсчитать абсолютное и относительное количество пропусков в каждом столбце.
  - Сделать промежуточный вывод: указать, для каких столбцов характерны пропуски, предположить причины их возникновения, и определить действия (замена, удаление или оставление).

- **Обработка дубликатов:**
  - Изучить уникальные значения в категориальных столбцах (жанр, платформа, рейтинг, год выпуска).
  - Нормализовать текстовые данные (привести к одному регистру, удалить лишние пробелы).
  - Найти и удалить явные дубликаты, посчитать удалённые строки.

- **Фильтрация данных:**
  - Отобрать данные по времени выхода игры: период с 2000 по 2013 год включительно.
  - Сохранить новый срез данных в отдельном датафрейме (например, `df_actual`).

- **Категоризация оценок:**
  - Для пользовательских оценок:
    - Высокая оценка: от 8 до 10 включительно.
    - Средняя оценка: от 3 до 8 (не включая 8).
    - Низкая оценка: от 0 до 3 (не включая 3).
  - Для оценок критиков:
    - Высокая оценка: от 80 до 100 включительно.
    - Средняя оценка: от 30 до 80 (не включая 80).
    - Низкая оценка: от 0 до 30 (не включая 30).
  - Добавить соответствующие столбцы с категориями в датафрейм.

- **Анализ платформ:**
  - Определить топ-7 платформ по количеству игр, выпущенных за указанный период.

- **Итоговый вывод:**
  - Составить общий вывод, описав проделанную работу:
    - Какие изменения внесены (столбцы переименованы, типы данных преобразованы, пропуски обработаны, дубликаты удалены).
    - Описание среза данных (`df_actual`) и новых полей (например, категории оценок).</font>

---

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Отлично по описанию. Цели и задачи ясны. Контекст проекта будет понятен любому потенциальному читателю.
 

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 В финальной версии подсказки и описания можно убрать. Наш проект − презентация работы и аргументации решений.  
(в последующих проектах, кажется, их уже будет либо меньше, либо не будет вообще)
 

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [4]:
import pandas as pd

#выгружаем данные в датафрейм
df = pd.read_csv('/datasets/new_games.csv')

#первые строки методом info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [1]:
import pandas as pd

#выгружаем данные в датафрейм
df = pd.read_csv('/datasets/new_games.csv')

#считаем число пропусков и их долю от столбца
total_rows = df.shape[0]
missing_values = df.isnull().sum()
missing_percentage = (df.isnull().mean() * 100).round(1)
zero_values = (df == 0).sum()
zero_percentage = ((df == 0).mean() * 100).round(1)

# Вывод результатов
missing_data = pd.DataFrame({'Всего записей': total_rows,'Пропуски': missing_values, 'Доля пропусков (%)': missing_percentage, 'Значения = 0': zero_values, 'Доля значений = 0 (%)': zero_percentage})
display(missing_data) #исправил вывод на display


Всего записей  Пропуски  Доля пропусков (%)  Значения = 0  \
Name                     16956         2                 0.0             0   
Platform                 16956         0                 0.0             0   
Year of Release          16956       275                 1.6             0   
Genre                    16956         2                 0.0             0   
NA sales                 16956         0                 0.0          4574   
EU sales                 16956         0                 0.0             0   
JP sales                 16956         0                 0.0             0   
Other sales              16956         0                 0.0          6701   
Critic Score             16956      8714                51.4             0   
User Score               16956      6804                40.1             0   
Rating                   16956      6871                40.5             0   

                 Доля значений = 0 (%)  
Name                               0.0  
Platform                           0.0  
Year of Release                    0.0  
Genre                              0.0  
NA sales                          27.0  
EU sales                           0.0  
JP sales                           0.0  
Other sales                       39.5  
Critic Score                       0.0  
User Score                         0.0  
Rating                             0.0


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь если выводить таблицу не через print(), а через display(), будет удобнее все воспринимать. 
</div>  


<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Cнова привет.  По исправлению отлично. 
</div> 

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

- **Размер датасета:** 16956 записей, 11 столбцов.
- **Типы данных:** 4 столбца с типом float64 и 7 столбцов с типом object.
- **Пропуски и их доля:**  
  - **Name:** 2 пропуска (0.0%)  
  - **Platform:** 0 пропусков (0.0%)  
  - **Year of Release:** 275 пропусков (1.6%)  
  - **Genre:** 2 пропуска (0.0%)  
  - **NA sales:** 0 пропусков (0.0%)  
  - **EU sales:** 0 пропусков (0.0%)  
  - **JP sales:** 0 пропусков (0.0%)  
  - **Other sales:** 0 пропусков (0.0%)  
  - **Critic Score:** 8714 пропусков (51.4%)  
  - **User Score:** 6804 пропусков (40.1%)  
  - **Rating:** 6871 пропусков (40.5%)  

- **Число нулевых значений и их доля:**  
  - **NA sales:** 4574 (27.0%)  
  - **Other sales:** 6701 (39.5%)  
  - **Остальные столбцы:** не содержат нулевых значений (0.0%)  

Особенности на которые можно обратить внимание: столбцы **EU sales**, **JP sales**, **User Score** имеют тип данных object. С учетом того, что в них содержатся количественные данные, возможно стоит изменить тип данных на числовой.


  
<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Здесь все верно по выявленным проблемам в данных. Можно также обратить внимание на несоответствия стилю названий колонок.
 

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [6]:
import pandas as pd

#выгружаем данные в датафрейм
df = pd.read_csv('/datasets/new_games.csv')

# Выводим названия всех столбцов
print("Названия столбцов:")
print(df.columns)

Названия столбцов:
Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


In [4]:
import pandas as pd

#выгружаем данные в датафрейм
df = pd.read_csv('/datasets/new_games.csv')

#приводим к стилю snake_case (приводим к нижнему регистру и заменяем пробелы на подчёркивания)
df.columns = (df.columns.str.lower().str.replace(' ', '_')) 

#вывод
print("\nНазвания столбцов в стиле snake_case:")
print(df.columns)


Названия столбцов в стиле snake_case:
Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Хорошее решение




### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

Столбцы **eu_sales**, **jp_sales**, **user_score** имеют тип данных object. С учетом того, что в них содержатся количественные данные, возможно стоит изменить тип данных на числовой. Причины почему тип данных может быть некорректен: ошибка в сборе данных (все 3 столбца), наличие строковых значений (в jp_sales *unknown*, в user_score *tbd* и *пустые значения (blanks)*) 

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [6]:
import pandas as pd

#число строк после загрузки 
total_initial_rows = df.shape[0]

#выгружаем данные в датафрейм
df = pd.read_csv('/datasets/new_games.csv')

#приводим названия столбцов к стилю snake_case (приводим к нижнему регистру и заменяем пробелы на подчёркивания)
df.columns = df.columns.str.lower().str.replace(' ', '_')

#преобразуем столбцы с числовыми значениями, содержащие текстовые данные, в float
columns_to_convert = ['jp_sales', 'eu_sales', 'user_score']

#заменяем 'unknown', 'tbd' и пустые строки на NaN
df[columns_to_convert] = df[columns_to_convert].replace(['unknown', 'tbd', ''], pd.NA)

#преобразуем столбцы к числовому типу данных
df[columns_to_convert] = df[columns_to_convert].apply(pd.to_numeric, errors='coerce')

#вывод
print(df[columns_to_convert].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   jp_sales    16952 non-null  float64
 1   eu_sales    16950 non-null  float64
 2   user_score  7688 non-null   float64
dtypes: float64(3)
memory usage: 397.5 KB
None


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Выше дублируешь приведение типов `df.columns = df.columns.str.lower().str.replace(' ', '_')` - в проектах стоит избегать дублирования кода.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.



- **Пропуски и их доля (посчитаны в разделе 1):**  
  - **name:** 2 пропуска (0.0%)  
  - **platform:** 0 пропусков (0.0%)  
  - **year_of_release:** 275 пропусков (1.6%)  
  - **genre:** 2 пропуска (0.0%)  
  - **na_sales:** 0 пропусков (0.0%)  
  - **eu_sales:** 0 пропусков (0.0%)  
  - **jp_sales:** 0 пропусков (0.0%)  
  - **other_sales:** 0 пропусков (0.0%)  
  - **critic_score:** 8714 пропусков (51.4%)  
  - **user_score:** 6804 пропусков (40.1%)  
  - **rating:** 6871 пропусков (40.5%)  

- **Число нулевых значений и их доля:**  
  - **na_sales:** 4574 (27.0%)  
  - **other_sales:** 6701 (39.5%)  
  - **Остальные столбцы:** не содержат нулевых значений (0.0%)  

In [25]:
#вывод рассчета количества и доли пропусков кодом.

#считаем число пропусков и их долю от столбца
total_rows = df.shape[0]
missing_values = df.isnull().sum()
missing_percentage = (df.isnull().mean() * 100).round(1)
zero_values = (df == 0).sum()
zero_percentage = ((df == 0).mean() * 100).round(1)

# Вывод результатов
missing_data = pd.DataFrame({'Всего записей': total_rows,'Пропуски': missing_values, 'Доля пропусков (%)': missing_percentage, 'Значения = 0': zero_values, 'Доля значений = 0 (%)': zero_percentage})
display(missing_data) #исправил вывод на display

Всего записей  Пропуски  Доля пропусков (%)  Значения = 0  \
name                     16679         0                 0.0             0   
platform                 16679         0                 0.0             0   
year_of_release          16679         0                 0.0             0   
genre                    16679         0                 0.0             0   
na_sales                 16679         0                 0.0          4513   
eu_sales                 16679         6                 0.0          5844   
jp_sales                 16679         4                 0.0         10461   
other_sales              16679         0                 0.0          6590   
critic_score             16679      1122                 6.7             0   
user_score               16679      1115                 6.7             1   
rating                   16679         0                 0.0             0   

                 Доля значений = 0 (%)  
name                               0.0  
platform                           0.0  
year_of_release                    0.0  
genre                              0.0  
na_sales                          27.1  
eu_sales                          35.0  
jp_sales                          62.7  
other_sales                       39.5  
critic_score                       0.0  
user_score                         0.0  
rating                             0.0

 <div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
При изменении вводых рассчеты могут поменяться.
   
Предлагаю здесь вывести рассчет количества и доли пропусков кодом.


<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 супер. Думаю, можно такую логику обернуть в функцию и использовать ее в последующих проектах.
</div> 

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


>**critic_score (51.4%), user_score (40.1%)** 
Пропуски могут быть связаны с тем, что не все игры получили оценки критиков и пользователей. 
Также возможна ошибка сбора данных, когда оценки просто не были внесены в систему или не были сохранены. В **user_score (40.1%)** Вместо оценок встречаются *tbd* и пустые значения (blanks). *tbd* (to be determined) означает, что на момент сбора данных оценки еще не были выставлены. Это значит, что убирать пропуски нельзя.
>>Действия по обработке пропусков: заполнить медианными значениями в разрезе платформы и жанра.

>**rating (40.5%)**. Пропуски в *rating* могут быть связаны с тем, что не все игры получили рейтинг ESRB. 

>>Действия по обработке пропусков: оставить как есть, заменив пропуски на unknown

>Нулевые значения в **na_sales (27.0%), other_sales (39.5%)**. Нули могут означать, что продажи в этих регионах были действительно нулевыми. Также возможна ошибка сбора данных, если информация о продажах отсутствовала. 
>>Действия по обработке пропусков - оставить, так как нули могут означать реальные нулевые продажи.  

	 
<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Возможно. Но возможно просто какие-то данные о продажах не попали в таблицу.

>**name, genre** - по два пропуска. 
>> Действия по обработке пропусков - удалить, так как слишком мало пропусков

>**year_of_release (1.6%)**. Причина пропусков может заключаться в том, что игра могла быть анонсирована, но не выпущена, как half life 3:)
>>Действия по обработке пропусков - удалить, так как доля мала, а поля мы не можем заполнить медианой или средним.

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 По аргументации и стратегии заполнения пропусков все замечательно.
</div>  


- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [26]:
#заполняем пропуски в 'critic_score' и 'user_score' медианными значениями в разрезе платформы и жанра
df['critic_score'] = df['critic_score'].fillna(df.groupby(['platform', 'genre'])['critic_score'].transform('median'))
df['user_score'] = df['user_score'].fillna(df.groupby(['platform', 'genre'])['user_score'].transform('median'))

In [27]:
#заполняем пропуски в rating значением 'unknown'
df['rating'] = df['rating'].fillna('unknown')

In [28]:
#удаляем строки без названия или жанра
df = df.dropna(subset=['name', 'genre'])

In [29]:
#удаляем строки с отсутствующим годом выпуска и приводим его к int
df = df.dropna(subset=['year_of_release'])
df['year_of_release'] = df['year_of_release'].astype(int)

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Супер. 
</div>  


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [30]:
#изучаем уникальные значения и выявляем неявные дубликаты
categorical_columns = ['genre', 'platform', 'rating', 'year_of_release']
for col in categorical_columns:
    print(f'Уникальные значения в столбце {col}:')
    print(df[col].sort_values().unique())
    print('-'*40)
print('Уникальные жанры:', df['genre'].unique())
print('Уникальные рейтинги:', df['rating'].unique())

Уникальные значения в столбце genre:
['action' 'adventure' 'fighting' 'misc' 'platform' 'puzzle' 'racing'
 'role-playing' 'shooter' 'simulation' 'sports' 'strategy']
----------------------------------------
Уникальные значения в столбце platform:
['2600' '3DO' '3DS' 'DC' 'DS' 'GB' 'GBA' 'GC' 'GEN' 'GG' 'N64' 'NES' 'NG'
 'PC' 'PCFX' 'PS' 'PS2' 'PS3' 'PS4' 'PSP' 'PSV' 'SAT' 'SCD' 'SNES' 'TG16'
 'WS' 'Wii' 'WiiU' 'X360' 'XB' 'XOne']
----------------------------------------
Уникальные значения в столбце rating:
['AO' 'E' 'E10+' 'EC' 'K-A' 'M' 'RP' 'T' 'UNKNOWN']
----------------------------------------
Уникальные значения в столбце year_of_release:
[1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993
 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007
 2008 2009 2010 2011 2012 2013 2014 2015 2016]
----------------------------------------
Уникальные жанры: ['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'figh

In [31]:
#нормализация текстовых данных
#приводим жанры и названия игр к нижнему регистру
df['genre'] = df['genre'].str.lower()
df['name'] = df['name'].str.lower()

#приводим рейтинг к верхнему регистру
df['rating'] = df['rating'].str.upper()

#проверка после нормализации
print('После нормализации:')
print('Уникальные жанры:', df['genre'].unique())
print('Уникальные рейтинги:', df['rating'].unique())

После нормализации:
Уникальные жанры: ['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']
Уникальные рейтинги: ['E' 'UNKNOWN' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']


 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь хитрая ситуация. У нас есть игры как из Европы, так и из Азии. В основном используется рейтинг ESRB https://ru.wikipedia.org/wiki/Entertainment_Software_Rating_Board 
    
Но некоторых категорий из представленных у нас в нем нет. Фактически, это сходные категории, но в другом обозначении.
    
Например, можно заменить значения 'К-А' на 'Е'.
    
Рекомендую в подобных случаях уточнять информацию в интернете (в реальных проектах на реальной работе также, ну и можно всегда спрашивать у коллег, особенно у тех, кто предоставил данные. Часто такие категории объясняются в паспорте данных. Ну в нашем случае нужно залезть в интернет). 


- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [33]:
#находим дубликаты
#сортируем датафрейм по всем столбцам 
df_sorted = df.sort_values(by=df.columns.tolist())

#находим дубликаты
duplicates = df_sorted[df_sorted.duplicated(keep=False)]

#выводим дубликаты
display(duplicates) #вывели данные через display

name platform  year_of_release  \
15191                    beyblade burst      3DS             2016   
15192                    beyblade burst      3DS             2016   
15301                 11eyes: crossover     X360             2009   
15302                 11eyes: crossover     X360             2009   
4860   18 wheeler: american pro trucker      PS2             2001   
...                                 ...      ...              ...   
2909   yu-gi-oh! the falsebound kingdom       GC             2002   
6695                      zoo resort 3d      3DS             2011   
6696                      zoo resort 3d      3DS             2011   
8156                 zumba fitness rush     X360             2012   
8157                 zumba fitness rush     X360             2012   

              genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
15191  role-playing      0.00      0.00      0.03         0.00          74.0   
15192  role-playing      0.00      0.00      0.03         0.00          74.0   
15301     adventure      0.00      0.00      0.02         0.00          63.0   
15302     adventure      0.00      0.00      0.02         0.00          63.0   
4860         racing      0.20      0.15      0.00         0.05          61.0   
...             ...       ...       ...       ...          ...           ...   
2909       strategy      0.49      0.13      0.07         0.02          71.0   
6695     simulation      0.11      0.09      0.03         0.02          71.0   
6696     simulation      0.11      0.09      0.03         0.02          71.0   
8156         sports      0.00      0.16      0.00         0.02          73.0   
8157         sports      0.00      0.16      0.00         0.02          73.0   

       user_score   rating  
15191        7.80  UNKNOWN  
15192        7.80  UNKNOWN  
15301        6.95  UNKNOWN  
15302        6.95  UNKNOWN  
4860         5.70        E  
...           ...      ...  
2909         8.50  UNKNOWN  
6695         7.40        E  
6696         7.40        E  
8156         6.20     E10+  
8157         6.20     E10+  

[458 rows x 11 columns]

 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

здесь также лучше вывести таблицу через display()

In [34]:
#удаляем дубликаты
#сохраняем количество строк до удаления дубликатов 
initial_row_count = df.shape[0]

#считам число явных дубликатов
duplicates_count = df.duplicated().sum()

#удаляем дубликаты
df_no_duplicates = df_sorted.drop_duplicates().reset_index(drop=True)

#сохраняем количество строк после удаления дубликатов
final_row_count = df_no_duplicates.shape[0]

#выводим результаты
print(f"Количество строк до удаления дубликатов: {initial_row_count}")
print(f'Число явных дубликатов: {duplicates_count}')
print(f"Количество строк после удаления дубликатов: {final_row_count}")

Количество строк до удаления дубликатов: 16679
Число явных дубликатов: 229
Количество строк после удаления дубликатов: 16450


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

В исходных данных столбца genre было 24 уникальных значения.
После нормализации (приведения к нижнему регистру) осталось 12 уникальных значений.
Обнаружено и устранено **12 неявных дубликатов**, которые были вызваны разницей в регистре написания жанров.

В дадатфрейме было найдено также **207 явных дубликатов**, которые были удалены

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [35]:
#считаем количество удалённых строк в абсолютном и относительном значениях
rows_removed_total = total_initial_rows - final_row_count
rows_removed_percent = (rows_removed_total / total_initial_rows) * 100

#вывод
print(f'Всего удалено строк: {rows_removed_total} ({rows_removed_percent:.2f}%)')

Всего удалено строк: 506 (2.98%)


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Всё верно по расчетам: отдельно смотрим долю дубликатов, отдельно считаем долю удаленных строк за всю предобработку.
</div>    



- После проведения предобработки данных напишите общий промежуточный вывод.

***Пропуски и их доля:***
* *name: 2 пропуска (0.0%)* Действие: удаление строк, так как доля мала, а заполнение невозможно.

* *platform: пропусков нет (0.0%)* Действие: нет необходимости обрабатывать.

* *year_of_release: 275 пропусков (1.6%)* Действие: удаление строк, так как доля небольшая, и корректно заполнить нельзя.

* *genre: 2 пропуска (0.0%)* Действие: удаление строк, так как доля мала и заполнение некорректно.

* *critic_score: 8714 пропусков (51.4%)* Действие: заполнение медианными значениями в разрезе платформы и жанра.

* *user_score: 6804 пропуска (40.1%)* Действие: заполнение медианными значениями в разрезе платформы и жанра.

* *rating: 6871 пропуск (40.5%)* Действие: заполнение значением unknown, так как многие игры не получили рейтинг ESRB.

***Нулевые значения и их доля:***
* *na_sales: 4574 нулевых значений (27.0%)* Действие: оставить без изменений, так как могут отражать реальные нулевые продажи.

* *other_sales: 6701 нулевое значение (39.5%)* Действие: оставить без изменений, так как могут отражать реальные нулевые продажи.

Остальные столбцы не содержат нулевых значений.

***Неявные дубликаты.***
В столбце genre исходно было 24 уникальных значения.
После нормализации к нижнему регистру осталось 12 уникальных значений.
Устранено 12 неявных дубликатов, вызванных разницей в регистре написания жанров.

***Явные дубликаты.*** После нормализации найдено и удалено 207 явных дубликатов (полностью идентичные строки).
Всего удалено строк: 207 (1.24%) от исходного датасета.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Отлично по предобработке.
</div>    



---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [37]:
#фильтруем данные по году выпуска игр (с 2000 по 2013 включительно)
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]

#создали копию среза для дальнейших действий
df_actual = df_actual.copy()

#выводим
display(df_actual)

name platform  \
0                                            wii sports      Wii   
2                                        mario kart wii      Wii   
3                                     wii sports resort      Wii   
6                                 new super mario bros.       DS   
7                                              wii play      Wii   
...                                                 ...      ...   
16947                     men in black ii: alien escape       GC   
16949                woody woodpecker in crazy castle 5      GBA   
16950  score international baja 1000: the official game      PS2   
16952                                  lma manager 2007     X360   
16954                                  spirits & spells      GBA   

       year_of_release     genre  na_sales  eu_sales  jp_sales  other_sales  \
0                 2006    sports     41.36     28.96      3.77         8.45   
2                 2008    racing     15.68     12.76      3.79         3.29   
3                 2009    sports     15.61     10.93      3.28         2.95   
6                 2006  platform     11.28      9.14      6.50         2.88   
7                 2006      misc     13.96      9.18      2.93         2.84   
...                ...       ...       ...       ...       ...          ...   
16947             2003   shooter      0.01      0.00      0.00         0.00   
16949             2002  platform      0.01      0.00      0.00         0.00   
16950             2008    racing      0.00      0.00      0.00         0.00   
16952             2006    sports      0.00      0.01      0.00         0.00   
16954             2003  platform      0.01      0.00      0.00         0.00   

       critic_score  user_score   rating  
0              76.0        8.00        E  
2              82.0        8.30        E  
3              80.0        8.00        E  
6              89.0        8.50        E  
7              58.0        6.60        E  
...             ...         ...      ...  
16947          73.0        7.90        T  
16949          71.0        8.00  UNKNOWN  
16950          67.5        7.85  UNKNOWN  
16952          73.0        6.60  UNKNOWN  
16954          71.0        8.00  UNKNOWN  

[12980 rows x 11 columns]

In [41]:
#число строк ДО фильтрации
initial_row_count = df.shape[0]

#убрана часть кода с фильтрацией по периоду

#число строк ПОСЛЕ фильтрации
filtered_row_count = df_actual.shape[0]

#сколько строк было ОТФИЛЬТРОВАНО
rows_removed = initial_row_count - filtered_row_count

#выводим результаты
print(f'Количество строк до фильтрации: {initial_row_count}')
print(f'Количество строк после фильтрации: {filtered_row_count}')
print(f'Всего отфильтровано строк: {rows_removed}')

Количество строк до фильтрации: 16679
Количество строк после фильтрации: 12980
Всего отфильтровано строк: 3699


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь также нужно убрать дублирование кода с фильтрацией по периоду.

</div> 


---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [42]:
#используем срез датафрейма для категоризации
df_actual = df_actual.copy()

#пользовательская функция для категоризации оценок пользователей
def categorize_user_score(score):
    if score >= 8:
        return 'высокая оценка'
    elif 3 <= score < 8:
        return 'средняя оценка'
    elif 0 <= score < 3:
        return 'низкая оценка'
    else:
        return 'нет оценки'

#добавление столбца
df_actual.loc[:, 'user_score_category'] = df_actual['user_score'].apply(categorize_user_score)

#вывод
display(df_actual[['name', 'user_score', 'user_score_category']])

name  user_score  \
0                                            wii sports        8.00   
2                                        mario kart wii        8.30   
3                                     wii sports resort        8.00   
6                                 new super mario bros.        8.50   
7                                              wii play        6.60   
...                                                 ...         ...   
16947                     men in black ii: alien escape        7.90   
16949                woody woodpecker in crazy castle 5        8.00   
16950  score international baja 1000: the official game        7.85   
16952                                  lma manager 2007        6.60   
16954                                  spirits & spells        8.00   

      user_score_category  
0          высокая оценка  
2          высокая оценка  
3          высокая оценка  
6          высокая оценка  
7          средняя оценка  
...                   ...  
16947      средняя оценка  
16949      высокая оценка  
16950      средняя оценка  
16952      средняя оценка  
16954      высокая оценка  

[12980 rows x 3 columns]

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [43]:
#пользовательская функция для категоризации оценок критиков
def categorize_critic_score(score):
    if score >= 80:
        return 'высокая оценка'
    elif 30 <= score < 80:
        return 'средняя оценка'
    elif 0 <= score < 30:
        return 'низкая оценка'
    else:
        return 'нет оценки'

#добавление столбца с категорией оценки критиков
df_actual.loc[:, 'critic_score_category'] = df_actual['critic_score'].apply(categorize_critic_score)

#вывод
display(df_actual[['name', 'critic_score', 'critic_score_category']])

name  critic_score  \
0                                            wii sports          76.0   
2                                        mario kart wii          82.0   
3                                     wii sports resort          80.0   
6                                 new super mario bros.          89.0   
7                                              wii play          58.0   
...                                                 ...           ...   
16947                     men in black ii: alien escape          73.0   
16949                woody woodpecker in crazy castle 5          71.0   
16950  score international baja 1000: the official game          67.5   
16952                                  lma manager 2007          73.0   
16954                                  spirits & spells          71.0   

      critic_score_category  
0            средняя оценка  
2            высокая оценка  
3            высокая оценка  
6            высокая оценка  
7            средняя оценка  
...                     ...  
16947        средняя оценка  
16949        средняя оценка  
16950        средняя оценка  
16952        средняя оценка  
16954        средняя оценка  

[12980 rows x 3 columns]

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [45]:
#считаем количество игр по категориям ПО оценкам пользователей
category_counts = df_actual['user_score_category'].value_counts()

#результат
print('Количество игр в каждой категории ПО оценкам пользователей:')
print(category_counts)

Количество игр в каждой категории ПО оценкам пользователей:
средняя оценка    9607
высокая оценка    3149
низкая оценка      119
нет оценки         105
Name: user_score_category, dtype: int64


In [46]:
#считаем количество игр по категориям ПО оценкам критиков
critic_category_counts = df_actual['critic_score_category'].value_counts()

#результат
print('Количество игр в каждой категории ПО оценкам критиков:')
print(critic_category_counts)

Количество игр в каждой категории ПО оценкам критиков:
средняя оценка    10974
высокая оценка     1843
нет оценки          108
низкая оценка        55
Name: critic_score_category, dtype: int64


- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [47]:
#считаем количество игр по платформам и выделение топ-7
top_7_platforms = df_actual['platform'].value_counts().head(7)

#вывод
print('Топ-7 платформ по количеству выпущенных игр за актуальный период (2000-2013 гг.):')
print(top_7_platforms)

Топ-7 платформ по количеству выпущенных игр за актуальный период (2000-2013 гг.):
PS2     2154
DS      2146
Wii     1294
PSP     1199
X360    1138
PS3     1107
GBA      826
Name: platform, dtype: int64


<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Здесь все корректно по всем рассчетам.
</div>  
	  

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

### Проверка ошибок в данных и их предобработка

- Приведены все названия к стилю snake_case (нижний регистр, пробелы заменены на подчёркивания).

### Обработка пропусков

* Удалены незначительные пропуски в столбцах name, genre, year_of_release.

* Значительные пропуски в оценках критиков (critic_score) и пользователей (user_score) заполнены медианными значениями, рассчитанными по сочетанию платформы и жанр.

* Пропуски в рейтинге ESRB (rating) заполнены значением unknown.

### Работа с дубликатами

* Были выявлены и устранены 12 неявных дубликатов в названиях жанров (за счёт нормализации к нижнему регистру).

* Удалены 207 явных дубликатов - полностью идентичных записей.

### Фильтрация данных

* Для коллег по запросу был сделан срез данных по периоду с 2000 по 2013 годы включительно, так как он представляет наибольший интерес для анализа трендов в игровой индустрии начала XXI века.

### Категоризация данных

- Добавлен новый столбец user_score_category, отражающий категорию оценки пользователей:
 - высокая оценка (от 8 до 10 включительно);
 - средняя оценка (от 3 до 8, не включая 8);
 - низкая оценка (от 0 до 3, не включая 3).

- Добавлен новый столбец critic_score_category, отражающий категорию оценки критиков:
 - высокая оценка (от 80 до 100 включительно);
 - средняя оценка (от 30 до 80, не включая 80);
 - низкая оценка (от 0 до 30, не включая 30).

### Анализ популярных платформ:

- Определены 7 платформ с наибольшим количеством выпущенных игр за выбранный период. Ими оказались: PS2, DS, Wi, PSP, X360, PS3, GBA 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Проект о предобработке, в общем выводе описываем, что делаем по предобработке. Все логично, отлично по оформлению.
</div>    




<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Внимательная работа, практически все решения верные и оптимальные, но кое-что нужно доработать и кое-что можно улучшить, чтобы проект стал совсем замечательным.
    
**Отмечу положительные моменты проекта🚀**
    
* Отличное оформление. Есть описание, комментарии, выводы.
* Хорошее владение кодом, инструментами предобработки. Все сложные моменты выполнены отлично.
* Хорошая аргументация решений. Во всех случаях понятно, почему принимали те или иные решения.
    
**На что стоит обратить внимание🤔**
    
* Важно вывести кодом количество пропусков в данных. Не стоит приводить рассчеты текстовым комментарием. 
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>




<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за доработку. Критичных комментариев нет. 
    
Если что-то пропустил или остались вопросы, напиши в телеграм @kriill

Желаю удачи в дальнейшей учебе!

</div>